python-checks/src/test/java/org/sonar/python/checks/BuiltinGenericsOverTypingModuleCheckTest.java

```diff
@@ -36,4 +40,139 @@ public void checkCollections() {
+    String codeWithIssue = code(
+      "from typing import Iterable",
+      "def foo(test: Iterable[int]) -> dict[str, int]:",
+      "    return {}");
+    String codeFixed = code(
+      "from typing import Iterable",
+      "def foo(test: collections.abc.Iterable[int]) -> dict[str, int]:",
+      "    return {}");
+    BuiltinGenericsOverTypingModuleCheck check = new BuiltinGenericsOverTypingModuleCheck();
+    PythonQuickFixVerifier.verify(check, codeWithIssue, codeFixed);
+    String specificMessage = String.format(MESSAGE, "collections.abc.Iterable");
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, codeWithIssue, specificMessage);
+  }
+
+  @Test
+  public void testQuickFixCollections() {
+    String codeWithIssue = code(
+      "from typing import Counter",
+      "def foo() -> Counter[str]:",
+      "    return None");
+    String codeFixed = code(
+      "from typing import Counter",
+      "def foo() -> collections.Counter[str]:",
+      "    return None");
+    BuiltinGenericsOverTypingModuleCheck check = new BuiltinGenericsOverTypingModuleCheck();
+    PythonQuickFixVerifier.verify(check, codeWithIssue, codeFixed);
+    String specificMessage = String.format(MESSAGE, "collections.Counter");
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, codeWithIssue, specificMessage);
+  }
+
+  @Test
+  public void testQuickFixNestedTypes() {
+    String codeWithIssue = code(
+      "from typing import Sequence",
+      "def foo() -> tuple[str, Sequence[int]]:",
+      "    return (\"foo\", list())");
+    String codeFixed = code(
+      "from typing import Sequence",
+      "def foo() -> tuple[str, collections.abc.Sequence[int]]:",
+      "    return (\"foo\", list())");
+    BuiltinGenericsOverTypingModuleCheck check = new BuiltinGenericsOverTypingModuleCheck();
+    PythonQuickFixVerifier.verify(check, codeWithIssue, codeFixed);
+    String specificMessage = String.format(MESSAGE, "collections.abc.Sequence");
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, codeWithIssue, specificMessage);
+  }
+
+  @Test
+  public void testQuickFixNestedFunction() {
+    String codeWithIssue = code(
+      "from typing import defaultdict",
+      "def foo() -> tuple[str, int]:",
+      "    def bar(test: defaultdict[str, int]) -> str:",
+      "        return \"bar\"",
+      "    return (\"foo\", list())");
+    String codeFixed = code(
+      "from typing import defaultdict",
+      "def foo() -> tuple[str, int]:",
+      "    def bar(test: collections.defaultdict[str, int]) -> str:",
+      "        return \"bar\"",
+      "    return (\"foo\", list())");
+    BuiltinGenericsOverTypingModuleCheck check = new BuiltinGenericsOverTypingModuleCheck();
+    PythonQuickFixVerifier.verify(check, codeWithIssue, codeFixed);
+    String specificMessage = String.format(MESSAGE, "collections.defaultdict");
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, codeWithIssue, specificMessage);
+  }
```

Comment:
It's a special case that came to my mind now, but could we test the case when we have an obsolete generic type without type arguments and that we do not raise an issue (and of course offer no quick fix):
```
def f() -> typing.List:
    pass
```

Reply:
Yes, good call I will add this to the main test.